Set handle size for sliders, and set consistent handle style

TryTwo opened the pull request:
Slider handles had no size set, so were super tiny.  Also set all sliders to have ticks below them, so the handles can all look the same. This look is nicer than square handles.

Novice at Qt, but I think I did it correctly..

JosJuice commented:
Could you post screenshots for comparison?

TryTwo commented:
![Comparison](https://i.imgur.com/HzCJzbv.png)

JosJuice commented:
That does look better! Though for some reason, the slider handles showed up fine on my system (Windows 10) even before this PR...

TryTwo commented:
That's odd. I even tried portable mode and it still looks like that.  I also can't seem to get a custom style to affect it, if that's any clue.

TryTwo commented:
Removed global tick size from GraphicsSlider and corrected the values for each Slider that uses it.  Fixed Game Config Slider.  Do we want to show the specific value for any more sliders? Like how CPU override does.

Techjar commented:
These commits should be squashed and the messages changed to match the PR.

TryTwo commented:
Added audio sliders.
I realized I'm actually setting the height for the entire slider element, and then Qt must be fixing the handle height from that.  I also figured out how to change the handles in the style sheets - you have to set handle.background, handle.height, and handle.margin for it to take the change. This still won't work if the height isn't defined (for me).

Sorry, still learning how to use github stuff. Messages- do you mean the name of the changes?

Techjar commented:
Commits don't have a "name", only a message. The message should be a description of the commit's changes just like PR itself.

TryTwo commented:
Fixed audio slider dynamically resizing as the digits in the value next to it changed. Also lined things up better. It seems tickdown style kind of makes the slider look off center when right next to text.

Can I rename the branch I'm PRing? I didn't know the name became a part of the PR itself. Could be more descriptive.

JosJuice commented:
No, the branch name of a PR can't be changed. But it doesn't need to be super descriptive, that's what the commit message is for. I think the current branch name is fine.

TryTwo commented:
I think this PR is done unless someone doesn't like my methods of setting a fixed height to widget elements .

lioncash commented:
You should likely be sizing elements based off screen metrics, rather than hardcoding fixed pixel heights/widths (which is explicitly advised against in [Qt's documentation on Hi-DPI displays](https://doc.qt.io/qt-5/highdpi.html)), otherwise the elements can show up too small on varying Hi-DPI monitor types.

TryTwo commented:
I found the issue of my QSlider being too small and not styled correctly (top pic) is due to some Windows 10 customization I did. Not sure exactly what customization. It may relate to BhaaLseN's referenced PR above. 

I couldn't find if QSize will resize settings for high DPI. Unfortunately I can't test it myself.  Maybe I can run something to get the widget height then feed that back into setheight and it'll fix the bug. I don't know much about sizing with screen metrics, if that's the only answer.

TryTwo commented:
I changed it to setMinimumHeight so it could still become larger if needed.  Other than that I'd need some help.  It really just needs to be set to the height of all the other layout elements, so it can't bug out and decide on 0px.

DXGLdotinfo commented:
For me, the skinny sliders occurred when I added a compatibility manifest to Dolphin.exe enabling themes in High Contrast Mode.  Whether or not my PR gets merged rests on this PR being implemented in a fully sane manner, as unthemed High Contrast Mode doesn't cause the issue.

TryTwo commented:
I managed to force VirtualBox to both 3840x2160 and 800x600 then have it rescale to my monitor. There were no issues with slider sizing when using either setminimumheight or setfixedheight.  

QSize also mentioned it uses "integer point precision" so these settings are not pixel based, and are expected to scale correctly.  I'm going to call DPI a non-issue. Let's commit this and in the unlikely event it makes sliders look weird on some platform, we can always fix it after hearing about the specifics.

lioncash commented:
setMinimumHeight is fine (in this instance), since it doesn't clamp the ability for the controls to grow.

I don't even see how QSize is relevant here, since it's just a struct that holds integer values. What the values for a QSize mean can vary depending on which particular API of the Qt libraries they're used with. They can still represent ranges of pixels. The only reason `integer-point precision` is used in it's description is to differentiate it from QSizeF, which uses floating-point precision.

TryTwo commented:
Changed all the sizes to be font based.  Set slider label's minimum width to be based on the max length of the text string used (ie  " 300 ms "), so it push the slider bars as the length changes.. If there's no problem getting the font with QFont::QFont() then I think this is a responsible fix.

TryTwo commented:
I rebased it. I also realized slider label misalignment can be fixed by specifically aligning all slider labels to AlignLeft or AlignRight, so I did that.

TryTwo commented:
I spent a good bit a time checking style sheets and found that adding:
```
QSlider {
    min-height: 20px;
    min-width: 20px;
}
```
to a stylesheet will correct the issue with slider bars. Should this be done by default?

DXGLdotinfo commented:
Are those style sheets corrected for HiDPI displays on Windows or is it 20 pixels regardless of DPI setting?

TryTwo commented:
To re-iterate: The bug is that the sizehint of sliders is broken for some people and sets the height to 3, which prevents the handle from being sized correctly.

The solution is to manually set the minimum height, the possible methods I found are::

- Set the minimum to 16-20 and hope it works for all platforms. This might be consistent with the stylesheet fix.
- Set the minimum to font height * 1.5, so it's tied to something (current)
- Set the minimum to the sizehint of another object that isn't bugged, like the combo box.
- Set the minimum in a stylesheet, which appears fine for High DPI stuff.  I don't know how to properly do this outside of custom user styles. Setting a custom userstyle breaks the white background on tabs for me (becomes grey).

jordan-woyak commented:
We've had a few Qt updates I think. Are sliders still bugged? Does this need a rebase?

TryTwo commented:
The bug is still there. I'm using a new windows installation which doesn't have it, so forgot about this.  My fix here also corrects universal janky behavior where the slider bars resize when the text label changes sizes. Some of my code probably needs to be updated.  If someone wants to go through the trouble of checking and approving it, I can update everything.

Tilka commented:
Hello from 3 years later. Is this still an issue?

TryTwo commented:
Yes. If there's interest in fixing it I can look back into it.  Check the audio buffer slider. As the number of digits change the slider length pops. I consider it janky.

I think one problem was testing it on multiple platforms.   I could look at our other methods to deal with widget sizing issues and see if there's a more current method.

Tilka commented:
>Check the audio buffer slider. As the number of digits change the slider length pops. I consider it janky.

Ah yes, I can see it now. I don't see the height issue from the initial screenshot though.


>I think one problem was testing it on multiple platforms.

I can test it on Linux/Mac/Windows if you tell me what else to look for.

TryTwo commented:
The height issue still exists on windows high contrast mode. Not sure about the latest windows versions though. Maybe on other uncommon setups too.

Places:
- Config -> Audio. 2x.  Don't know how to enable the "decoding quality" slider.
    *Buffer Size slider has the numeric label min width fixed.
- Config -> Wii. 2x.
- Config ->Advanced .3x
    *CPU clock has the min width fixed.
- Graphics -> Enhancements. 2x
- Graphics -> Hacks. 1x.
- Game Properties -> Game Config. 1x.

TryTwo commented:
Color correction sliders included.  Set the third slider onto a new line, because the original space for it gave it poor granularity and this also fixes the popping due to the label changing size.

TryTwo commented:
When using high contrast in Windows, the normal themes are still broken and the Fusion themes work correctly. Presumably because fusion re-styles the sliders. Perhaps intentionally defining the slider handles in all themes would work, but I'm not sure about cross-platform styling.

TryTwo commented:
Thanks for looking this over. I agree with your points and separating it out. I'm still not sure what the best fix is either, but at least we have fusion styles now.
